### PR TITLE
Typed analyzers

### PR DESCRIPTION
**Pull request by skystrife**
This addresses issue #106.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -237,6 +237,27 @@ if (META_HAS_NONEMPTY_HASH_SUPPORT)
                              -DMETA_HAS_NONEMPTY_HASH_SUPPORT)
 endif()
 
+# work around a bug in gcc 4.8.1 where having a lambda as a default
+# argument to a member function template causes an internal compiler error
+check_cxx_source_compiles("
+struct h {
+  template <class Function = void(*)()>
+  void foo(Function&& fun = [](){}) {
+    fun();
```
Comment (smassung) [inline CMakeLists.txt]: [](){}

File: include/analyzers/analyzer.h [hunk 2]
```diff
@@ -48,71 +69,88 @@ class analyzer
...
 
-    /**
-     * @param src The token stream that will feed into this filter
-     * @param config The config group used to create the filter from
-     * @return a single filter specified by a config object
-     */
-    static std::unique_ptr<token_stream>
-        load_filter(std::unique_ptr<token_stream> src,
-                    const cpptoml::table& config);
+extern template std::unique_ptr<analyzer<uint64_t>>
```
Comment (smassung) [inline include/analyzers/analyzer.h]: Can you explain what the `extern`s accomplish here? (For documentation)

File: src/analyzers/ngram/ngram_word_analyzer.cpp [hunk 3]
```diff
@@ -16,52 +16,61 @@ namespace meta
 namespace analyzers
 {
 
-const std::string ngram_word_analyzer::id = "ngram-word";
+template <class T>
+const std::string ngram_word_analyzer<T>::id = "ngram-word";
```
Comment (smassung) [inline src/analyzers/ngram/ngram_word_analyzer.cpp]: We can probably create just `ngram_analyzer` and have it use different input streams, for example unigram words or POS tags. Not for this PR, but in general.

**Comment by smassung** [comment]
We're going to have a lot of documentation rewriting on meta-toolkit.org for 2.0...

File: include/analyzers/analyzer.h (hunk 2 above)
Comment (skystrife) [inline include/analyzers/analyzer.h]: So these show up all over the place now. Basically every `analyzer` header file and every `tree_featurizer` header file. Do you want me to document it everywhere, or just here?

Essentially what this is doing is telling the compiler to not attempt to instantiate the template for the specified type in all translation units that include this header. This would obviously cause an error at link time unless you link with an object file that actually has the instantiation. We force the compiler to instantiate the templates *once* in the corresponding `.cpp` file. This is what saves us from having to move everything to be header-only and parse and instantiate every function that's used in every translation unit.

Hmm... I guess an easier explanation is that this:

```cpp
extern template class myanalyzer<uint64_t>;
```

is the declaration and 

```cpp
template class myanalyzer<uint64_t>;
```

is the definition, to use the standard terms for regular classes. See: http://en.cppreference.com/w/cpp/language/class_template#Explicit_instantiation

**tl;dr:** Not sure how you want me to document what this is doing, beyond something like


```cpp
// declare the valid instantiations for this analyzer
extern template class ngram_word_analyzer<uint64_t>;
extern template class ngram_word_analyzer<double>;

// declare the valid instantiations for this analyzer's trait class
extern template struct analyzer_traits<ngram_word_analyzer<uint64_t>>;
extern template struct analyzer_traits<ngram_word_analyzer<double>>;
```

Comment (smassung) [inline include/analyzers/analyzer.h]: I think the one-liner comments are good since this would seem a little unusual for someone that happens to see this.

**Comment by smassung** [comment]
I'm going to merge this; the travis failures are spurious.
